pylot/planning/frenet_optimal_trajectory/FrenetOptimalTrajectory/src/FrenetOptimalTrajectory/FrenetOptimalTrajectory.h

```diff
@@ -0,0 +1,44 @@
+// Author: Edward Fang
+// Email: edward.fang@berkeley.edu
+//
+// This code is adapted from https://github.com/AtsushiSakai/PythonRobotics/tree/
+// master/PathPlanning/FrenetOptimalTrajectory.
+// Its author is Atsushi Sakai.
+//
+// Reference Papers:
+// - [Optimal Trajectory Generation for Dynamic Street Scenarios in a Frenet Frame]
+// (https://www.researchgate.net/profile/Moritz_Werling/publication/224156269_Optimal_Trajectory_Generation_for_Dynamic_Street_Scenarios_in_a_Frenet_Frame/links/54f749df0cf210398e9277af.pdf)
+// - [Optimal trajectory generation for dynamic street scenarios in a Frenet Frame]
+// (https://www.youtube.com/watch?v=Cj6tAQe7UCY)
+
+#ifndef FRENETOPTIMALTRAJECTORY_FRENETOPTIMALTRAJECTORY_H
+#define FRENETOPTIMALTRAJECTORY_FRENETOPTIMALTRAJECTORY_H
+
+#include <vector>
+#include "../CubicSpline/CubicSpline2D.h"
+#include "FrenetPath.h"
+
+// The FrenetOptimalTrajectory class.
+// See the references above for details.
+class FrenetOptimalTrajectory {
+public:
+    CubicSpline2D csp;
+    std::vector<std::tuple<double, double>> obstacles;
+    std::vector<double> x, y;
+    double s0, c_speed, c_d, c_d_d, c_d_dd, target_speed;
+    FrenetPath best_frenet_path;
+    std::vector<FrenetPath*> frenet_paths;
+    std::vector<FrenetPath*> result_frenet_paths;
+    FrenetOptimalTrajectory(std::vector<double>& x, std::vector<double>& y,
+            double s0, double c_speed, double c_d, double c_d_d, double
+            c_d_dd, double target_speed,
+            std::vector<std::tuple<double, double>>& obstacles);
+private:
+    void calc_frenet_paths();
+    void calc_global_paths();
+    void validate_paths();
```

Comment:
I've noticed that each method does a pass over all computed paths. Wouldn't be nicer to have a compute_best_path method, which for each computed path, invoked a method that calculates a global path out of it, and another one that validates it. Maybe I'm missing something out, but to me it looks like the FrenetPath should have a to_global_path() method, and a is_colliding_with(obstacles) method. Let me know what you think.

Reply:
Agreed, done!